R/slide.R

```diff
@@ -373,3 +374,368 @@ epi_slide <- function(x, f, ..., before, after, ref_time_values,
+  }
+
+  pad_early_dates <- c()
+  pad_late_dates <- c()
+
+  # If dates are one of tsibble-provided classes, can step by numeric. `tsibble`
+  # defines a step of 1 to automatically be the quantum (smallest resolvable
+  # unit) of the date class. For example, one step = 1 quarter for `yearquarter`.
+  #
+  # `tsibble` classes apparently can't be added to in different units, so even
+  # if `time_step` is provided by the user, use a unit step.
+  if (inherits(x$time_value, c("yearquarter", "yearweek", "yearmonth")) ||
+    is.numeric(x$time_value)) {
+    all_dates <- seq(min(x$time_value), max(x$time_value), by = 1L)
+
+    if (before != 0) {
+      pad_early_dates <- Start(all_dates) - before:1
+    }
+    if (after != 0) {
+      pad_late_dates <- End(all_dates) + 1:after
+    }
+  } else if (missing(time_step)) {
+    # Guess what `by` should be based on the epi_df's `time_type`.
+    ttype <- attributes(x)$metadata$time_type
+    by <- switch(ttype,
+      day = "days",
+      week = "weeks",
+      yearweek = "weeks",
+      yearmonth = "months",
+      yearquarter = "quarters",
+      year = "years",
+      NA # default value for "custom", "day-time"
+    )
+
+    if (is.na(by)) {
+      Abort(
+        c(
+          "`frollmean` requires a full window to compute a result, but
+          `time_type` associated with the epi_df was not mappable to period
+          type valid for creating a date sequence.",
+          "i" = c("The input data's `time_type` was probably `custom` or `day-time`.
+          These require also passing a `time_step` function.")
+        ),
+        class = "epiprocess__epi_slide_mean__unmappable_time_type",
+        epiprocess__time_type = ttype
+      )
+    }
+
+    # Time_step can be any of `c("days", "weeks", "months", "quarters", "years")`
+    all_dates <- seq(min(x$time_value), max(x$time_value), by = by)
+
+    if (before != 0) {
+      pad_early_dates <- Start(all_dates) - before:1
+    }
+    if (after != 0) {
+      pad_late_dates <- End(all_dates) + 1:after
+    }
+  } else {
+    # A custom time step is specified
+
+    # Calculate the number of `time_step`s required to go between min and max time
+    # values. This is roundabout because difftime objects, lubridate::period objects,
+    # and Dates are hard to convert to the same time scale and add.
+    t_elapsed_s <- difftime(max(x$time_value), min(x$time_value), units = "secs")
+    step_size_s <- lubridate::as.period(time_step(1), unit = "secs")
+    n_steps <- ceiling(as.numeric(t_elapsed_s) / as.numeric(step_size_s))
+
+    all_dates <- min(x$time_value) + time_step(0:n_steps)
+
+    if (before != 0) {
+      pad_early_dates <- Start(all_dates) - time_step(before:1)
+    }
+    if (after != 0) {
+      pad_late_dates <- End(all_dates) + time_step(1:after)
+    }
+  }
+
+
+  # `frollmean` is 1-indexed, so create a new window width based on our
+  # `before` and `after` params.
+  m <- before + after + 1L
+
+  if (is.null(names_sep)) {
```

Comment:
Should the default naming be something mean-specific rather than `slide_value_{x}`?  Hard to think of a very satisfying thing, but still something with `mean` or `av` in it feels better maybe.

Reply:
Just left this as `slide_value` in https://github.com/cmu-delphi/epiprocess/pull/433.